<a href="https://colab.research.google.com/github/YaninaK/sales-forecast/blob/main/notebooks/04_Transformer_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sales forecast for 20 stores for 31 days' period.

## Transformer model

[EDA: omissions, clustering & Fourier transformations](
https://colab.research.google.com/drive/1udCgyEt7lusSY43lnduVDEnZvdvVEt9h?usp=sharing)

[Feature extraction](https://colab.research.google.com/drive/1H-vQoHnycp62YuZD1xrmIDjufpDu7ig3?usp=sharing)

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!git clone -q https://github.com/YaninaK/sales-forecast.git
!pip install -r sales-forecast/requirements_Colab.txt -q 

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 358.2/358.2 KB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 KB 8.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 136.2/136.2 KB 12.5 MB/s eta 0:00:00


In [3]:
%cd sales-forecast

/content/sales-forecast


In [4]:
import sys
import os

sys.path.append(os.getcwd())
sys.path.append(os.path.join(os.getcwd(), "src", "sales_forecast"))

In [5]:
import random
import pandas as pd
import numpy as np
import tensorflow as tf
import joblib

from data.make_dataset import load_data
from src.sales_forecast.models import train

from data.johnson_su_transformation import JohnsonSU
from data.train_test_datasets import get_train_dataset, get_test_dataset
from models.model_TF import get_model_TF

In [6]:
import warnings
warnings.filterwarnings('ignore')

In [7]:
PATH = '/content/drive/MyDrive/ML_projects/01_Time_series/sales_forecast/'
DEFAULT_RANDOM_SEED = 3

In [8]:
def set_all_seeds(seed=DEFAULT_RANDOM_SEED):
  random.seed(seed)
  os.environ['PYTHONHASHSEED'] = str(seed)
  np.random.seed(seed)
  tf.random.set_seed(seed)
  
set_all_seeds()

## 1. Data ingestion

In [9]:
data = load_data(PATH)

print(f'data.shape = {data.shape}\n')
data.head(2)

data.shape = (8230, 3)



id          dt  target
0   0  2016-01-02   747.8
1   0  2016-01-03   681.2

## 2. Data preprocessing pipeline

In [10]:
regenerate = True
if regenerate:
  X = train.data_preprocessing_pipeline(data)
else:
  X = np.load('data/04_feature/train_dataset.npy')

X.shape

Number of outliers = 8


(153, 20, 40)

## 3. Transformer model

### 3.1 Train dataset

In [11]:
X_train, y_train = get_train_dataset(X)
X_train.shape, y_train.shape

((900, 54, 40), (900, 27, 40))

### 3.2 Training of the model

In [12]:
model = get_model_TF(X_train.shape[1:])
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 54, 40)]     0           []                               
                                                                                                  
 layer_normalization (LayerNorm  (None, 54, 40)      80          ['input_1[0][0]']                
 alization)                                                                                       
                                                                                                  
 transformer_block (Transformer  (None, 54, 40)      43846       ['layer_normalization[0][0]']    
 Block)                                                                                           
                                                                                              

In [13]:
n_shops = 20
n_epochs = 20
batch_size = n_shops * 4
m = n_shops * 4

reduce_lr = tf.keras.callbacks.LearningRateScheduler(
    lambda epoch: 3e-2 * 0.95 ** epoch
)
model.compile(
    optimizer=tf.keras.optimizers.Adam(),
    loss=tf.keras.losses.Huber()
)
history = model.fit(
    X_train[:-m, :, :], y_train[:-m, :, :], 
    epochs=n_epochs,
    validation_data=(X_train[-m:, :, :], y_train[-m:, :, :]),
    batch_size=batch_size, 
    verbose=1,
    callbacks=[reduce_lr],
    shuffle=True,
    workers=-1,
    use_multiprocessing=True
)

Epoch 1/20
11/11 [==============================] - 58s 1s/step - loss: 0.8359 - val_loss: 1874.0446 - lr: 0.0300
Epoch 2/20
11/11 [==============================] - 10s 934ms/step - loss: 0.6345 - val_loss: 13.3830 - lr: 0.0285
Epoch 3/20
11/11 [==============================] - 13s 1s/step - loss: 0.5506 - val_loss: 1.4645 - lr: 0.0271
Epoch 4/20
11/11 [==============================] - 13s 1s/step - loss: 0.4867 - val_loss: 0.5317 - lr: 0.0257
Epoch 5/20
11/11 [==============================] - 13s 1s/step - loss: 0.4498 - val_loss: 0.4886 - lr: 0.0244
Epoch 6/20
11/11 [==============================] - 13s 1s/step - loss: 0.4350 - val_loss: 0.5439 - lr: 0.0232
Epoch 7/20
11/11 [==============================] - 12s 1s/step - loss: 0.4135 - val_loss: 0.4810 - lr: 0.0221
Epoch 8/20
11/11 [==============================] - 11s 958ms/step - loss: 0.4018 - val_loss: 0.4889 - lr: 0.0210
Epoch 9/20
11/11 [==============================] - 12s 1s/step - loss: 0.3918 - val_loss: 0.4698 - lr

### 3.3 Inference

In [14]:
X_test, y_test = get_test_dataset(X)

In [15]:
pred = model.predict(X_test)
pred.shape

1/1 [==============================] - 3s 3s/step


(20, 27, 40)

### 3.4 Model evaluation

#### Mean absolute scaled error (MASE) 
https://en.wikipedia.org/wiki/Mean_absolute_scaled_error

The denominator of mean absolute scaled error is the mean absolute error of the one-step "seasonal naive forecast method" on the training set, which uses the actual value from the prior 30 days as the forecast: 
* F(t) = Y(t−30)

MASE is scale-free error metric. So, it can be applied to scaled data without extra iteration of inverse scaling.

In [16]:
pred.shape, y_test.shape

((20, 27, 40), (20, 27, 40))

In [17]:
forecast_error = abs(pred[:, :, 0] - y_test[:, :, 0]).mean()
mase = forecast_error / abs(y_test[:, :, 6]).mean()

print(f'Mean absolute scaled error = {round(mase.mean(), 4)}')

Mean absolute scaled error = 0.8705


__Findings:__

* Out of 72 samples available for training (153 less input and output sequence length) 4 were used for validation of the model and 27 for testing. So, we had only 41 samples for training of the model and it is definately not sufficient.

* To get more accurate predictions, we need to retrain the model on the full set of data. It should improve the final results.

## 4. Final prediction

### 4.1. Dataset

In [18]:
X_train, y_train = get_train_dataset(X, final=True)
X_test = get_test_dataset(X, final=True)

X_train.shape, y_train.shape, X_test.shape

((1440, 54, 40), (1440, 27, 40), (20, 54, 40))

### 4.2. Retraining the model on full dataset

In [19]:
n_shops = 20
n_epochs = 20
batch_size = n_shops * 2
m = n_shops * 4

model = model = get_model_TF(X_train.shape[1:])
reduce_lr = tf.keras.callbacks.LearningRateScheduler(
    lambda epoch: 3e-2 * 0.95 ** epoch
)
model.compile(
    optimizer=tf.keras.optimizers.Adam(), 
    loss=tf.keras.losses.Huber()
)
history = model.fit(
    X_train[:-m, :, :], y_train[:-m, :, :], 
    epochs=n_epochs,
    validation_data=(X_train[-m:, :, :], y_train[-m:, :, :]),
    batch_size=batch_size, 
    verbose=1,
    callbacks=[reduce_lr],
    shuffle=True,
    workers=-1,
    use_multiprocessing=True,    
)

Epoch 1/20
34/34 [==============================] - 59s 668ms/step - loss: 0.6625 - val_loss: 30.9424 - lr: 0.0300
Epoch 2/20
34/34 [==============================] - 21s 621ms/step - loss: 0.4364 - val_loss: 0.5366 - lr: 0.0285
Epoch 3/20
34/34 [==============================] - 20s 571ms/step - loss: 0.3728 - val_loss: 0.4432 - lr: 0.0271
Epoch 4/20
34/34 [==============================] - 19s 570ms/step - loss: 0.3609 - val_loss: 0.3674 - lr: 0.0257
Epoch 5/20
34/34 [==============================] - 21s 624ms/step - loss: 0.3552 - val_loss: 0.3584 - lr: 0.0244
Epoch 6/20
34/34 [==============================] - 19s 572ms/step - loss: 0.3519 - val_loss: 0.3514 - lr: 0.0232
Epoch 7/20
34/34 [==============================] - 19s 567ms/step - loss: 0.3499 - val_loss: 0.3493 - lr: 0.0221
Epoch 8/20
34/34 [==============================] - 21s 617ms/step - loss: 0.3481 - val_loss: 0.3476 - lr: 0.0210
Epoch 9/20
34/34 [==============================] - 19s 569ms/step - loss: 0.3462 - val

### 4.3 Model evaluation on validation set

In [20]:
pred = model.predict(X_train[-m:, :, :])
pred[:, :, 0].shape,

3/3 [==============================] - 3s 111ms/step


((80, 27),)

In [21]:
forecast_error = abs(pred[:, :, 0] -  y_train[-m:, :, 0]).mean()
mase = forecast_error / abs( y_train[-m:, :, 6]).mean()

print(f'Mean absolute scaled error = {round(mase.mean(), 4)}')

Mean absolute scaled error = 0.7595


## 5. Findings

Transformer model has not outperformed LSTM model. There is not enouph data for transformers. So, for this particular task LSTM model is more appropriate.